alphadia/workflow/peptidecentric.py

```diff
@@ -444,89 +484,151 @@ def calibration(self):
             )
             return
 
-        self.start_of_calibration()
-        for current_epoch in range(self.config["calibration"]["max_epochs"]):
-            if self.check_epoch_conditions():
-                pass
-            else:
-                break
+        # Get the order of optimization
+        order_of_optimization = self.get_optimizers()
+
+        self.reporter.log_string(
+            "Starting initial classifier training and precursor identification.",
+            verbosity="progress",
+        )
+
+        # Get the optimization lock
+        eg_idxes_for_calibration, precursor_df, fragments_df = (
+            self.get_optimization_lock()
+        )
+
+        self.optimization_manager.fit(
+            {"classifier_version": self.fdr_manager.current_version}
+        )
+
+        # Perform a first recalibration on the optimization lock.
+        precursor_df_filtered, fragments_df_filtered = self.filter_dfs(
+            precursor_df, fragments_df
+        )
+
+        self.recalibration(precursor_df_filtered, fragments_df_filtered)
+
+        self.reporter.log_string(
+            "Required number of precursors found and required number of training iterations performed. Starting search parameter optimization.",
+            verbosity="progress",
+        )
 
-            self.start_of_epoch(current_epoch)
+        self.reporter.log_string(
+            "=== Performing initial optimization on extracted data. ===",
+            verbosity="info",
+        )
 
-            features = []
-            fragments = []
-            for current_step, (start_index, stop_index) in enumerate(self.batch_plan):
-                self.start_of_step(current_step, start_index, stop_index)
+        for optimizers in order_of_optimization:
+            for optimizer in optimizers:
+                optimizer.step(precursor_df_filtered, fragments_df_filtered)
+        # End of first recalibration
 
-                eg_idxes = self.elution_group_order[start_index:stop_index]
+        # Start of optimization/recalibration loop
+        for optimizers in order_of_optimization:
```

Comment:
Introducing a `first_recalibration_and_optimization()` method makes sense to me, as it is self-contained and somewhat distinct from the logic of the rest of calibration. But since a `recalibration_loop()` method would be more than half as long as the current `calibration()` method though, I'm not sure it would make things much clearer.